homeassistant/components/bluetooth/websocket_api.py

```diff
@@ -0,0 +1,159 @@
+    """Class to hold and manage the subscription data."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        connection: websocket_api.ActiveConnection,
+        ws_msg_id: int,
+        match_dict: BluetoothCallbackMatcher,
+    ) -> None:
+        """Initialize the subscription data."""
+        self.hass = hass
+        self.match_dict = match_dict
+        self.pending_service_infos: list[BluetoothServiceInfoBleak] = []
+        self.ws_msg_id = ws_msg_id
+        self.connection = connection
+        self.pending = True
+        # Keep time_diff precise to 2 decimal places
+        # so the cached serialization can be reused,
+        # however we still want to calculate it each
+        # subscription in case the system clock is wrong
+        # and gets corrected.
+        self.time_diff = round(time.time() - time.monotonic(), 2)
+
+    @callback
+    def _async_unsubscribe(
+        self, cancel_callbacks: tuple[Callable[[], None], ...]
+    ) -> None:
+        """Unsubscribe the callback."""
+        for cancel_callback in cancel_callbacks:
+            cancel_callback()
+
+    @callback
+    def async_start(self) -> None:
+        """Start the subscription."""
+        connection = self.connection
+        cancel_adv_callback = async_register_callback(
+            self.hass, self, self.match_dict, BluetoothScanningMode.PASSIVE
+        )
+        cancel_disappeared_callback = _get_manager(
+            self.hass
+        ).async_register_disappeared_callback(self._async_removed)
+        connection.subscriptions[self.ws_msg_id] = partial(
+            self._async_unsubscribe, (cancel_adv_callback, cancel_disappeared_callback)
+        )
+        self.pending = False
+        self.connection.send_message(
+            json_bytes(websocket_api.result_message(self.ws_msg_id))
+        )
+        self._async_added(self.pending_service_infos)
+        self.pending_service_infos.clear()
+
+    def _async_added(self, service_infos: Iterable[BluetoothServiceInfoBleak]) -> None:
+        self.connection.send_message(
+            json_bytes(
+                websocket_api.event_message(
+                    self.ws_msg_id,
+                    {
+                        "add": [
+                            serialize_service_info(service_info, self.time_diff)
+                            for service_info in service_infos
+                        ]
+                    },
+                )
+            )
+        )
+
+    def _async_removed(self, address: str) -> None:
+        self.connection.send_message(
+            json_bytes(
+                websocket_api.event_message(
+                    self.ws_msg_id,
+                    {
+                        "remove": [
+                            {
+                                "address": address,
+                            }
+                        ]
+                    },
+                )
+            )
+        )
+
+    def __call__(
```

Comment:
When is this used?

Reply:
<https://github.com/home-assistant/core/pull/134291/files#diff-d7419b0665ab475160c16f41bad2820afc82d78b88d1b2e2658c796c02ef926aR90>

The callback it `self`. I'll make it a function as it will be easier to read

Reply:
634c3ea8b9a